### ISSUE TITLE
[Performance] Reduce simulation runtime

### ISSUE THREAD

**Issue by jannikgro** [hpi-epic/BP2021#201]
The simulation speed gets a problem when Actor Critic (#60) is used. Consider the following optimizations in sim_market:

- [x] Try to avoid deepcopy. I think it is necessary right now, because output_dict is an instance variable. if you just return it, it is a reference. Changing it in a different file would corrupt its content. Try to better understand this problem and find a solution. One possibility: Make output_dict just a temporary variable in the function. If you return it, the reference will be deleted in the market and no problem will occur.

- [x] observation and generate_customer_offer np.concatenate is the main performance issue, because it creates new np arrays all the time. Just create one array at the beginning and insert there. (How it was done: generate_customer_offer was removed and observation was done with only one concatenation.)

- [x] simulate customers and owners: after the proabability distribution is generated once, try to run the shuffle_from_probability runs in parallel. Check, if numpy (or pytorch) offers a library method for fast tensor shuffling and use it if possible.

- [x] refactoring: You can consider owners and customers as part of the market, so you don't have to generate a np-array to communicate with them. Furthermore, that can increase the understandability and maintainabiliy significantly. It was remarked several times in the past. But it is an invasive refactoring and should be extracted as a new issue.

- [ ] Check if the way we handle the logging events right now is a performance issue. Currently, we have nested dictionaries for that purpose, but maybe we should just use one dictionary with numpy arrays inside. This would be an extra story too, because several files would be affected. (This last item was not done. It would require further commitment)
